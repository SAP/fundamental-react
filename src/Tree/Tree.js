import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import TreeNode from './_TreeNode';
import useUniqueId from '../utils/useUniqueId';
import withStyles from '../utils/withStyles';
import React, { useState } from 'react';
import styles from 'fundamental-styles/dist/tree.css';

const classnames = classnamesBind.bind(styles);

/**
 *
 *
 * Unlike tree tables, trees are used for rather basic data.
 *
 * USE THE TREE IF:
 *
 * * You need to display the key identifier of hierarchically structured items
 * * Selecting one or more items out of a set of hierarchically structured items is a main use case.
 * * The hierarchy has a restricted number of levels (up to about 12, depending on the content) and items (around 200).
 * * You want to have only one implementation for all devices.
 *
 * DO NOT USE THE TREE IF:
 *
 * * The main use case is to select one item from a very small number of non-hierarchical items,
 * without viewing additional details. In this case, a <a href="/?path=/docs/component-api-select" target="_blank">Select</a> or
 *  <a href="/?path=/docs/component-api-comboboxinput" target="_blank">ComboboxInput</a> might be more appropriate.
 * * Items are not structured hierarchically. Use a <a href="/?path=/docs/component-api-list" target="_blank">List</a> instead.
 * * The hierarchy turns out to have only two levels. In this case, use a grouped list.
 * * The hierarchy turns out to be just a categorization based on several details of the item.
 * In this case, an analytical table provides multi-level grouping. Note that the analytical table is not fully responsive.
 * It is only available for desktops and tablets, so you will need to take an adaptive approach by offering an additional UI for smartphones.
 * * You need to display very deep hierarchies with additional data per item. In this case, use a tree table (not available yet in Fundamental-styles).
 * Note that the tree table is not fully responsive. It is only available for desktops and tablets,
 * so you will need to take an adaptive approach by offering an additional UI for smartphones.
 * * The structure contains more than around 200 items. In this case, use the tree table (not available yet in Fundamental-styles).
 * It is optimized for large item sets and provides better performance. Note that the tree table is not fully responsive.
 * It is only available for desktops and tablets, so you will need to take an adaptive approach by offering an additional UI for smartphones.
 * * You need an overview of a large amount of data. In this case, use a chart.
*/
const Tree = React.forwardRef(({
    active,
    children,
    className,
    compact,
    cssNamespace,
    id,
    noBorders,
    emptyText,
    selection,
    selectionPosition,
    ...props
}, ref) => {
    const [expansionDepth, setExpansionDepth] = useState(0);
    const [tree, setTree] = useState({
        'root': {
            id: 'root',
            expanded: true,
            level: 0,
            children: {}
        }
    });

    const updateTree = ({
        toggledAtLevel,
        lastLevelExpanding,
        parentPath }) => {
        const pathArr = parentPath && parentPath.split && parentPath.split('/') || [];
        let checkTree = tree;
        let checkTreeRoot = checkTree;
        pathArr.forEach((pathItem, index) => {
            const currentItem = checkTree[pathItem];

            if (currentItem && index === pathArr.length - 1) {
                currentItem.expanded = lastLevelExpanding;
            }
            if (!currentItem) {
                checkTree[pathItem] = {
                    id: pathItem,
                    expanded: lastLevelExpanding,
                    level: toggledAtLevel,
                    children: {}
                };
                checkTree = checkTree[pathItem].children;
            } else {
                checkTree = currentItem && currentItem.children;
            }
        });
        setTree(checkTreeRoot);

    };

    const getExpandedChildren = (treeNode) => {
        if (!treeNode || !treeNode.children || Object.keys(treeNode.children).length === 0) return 0;
        const childrenIds = Object.keys(treeNode.children);
        const expandedChildrenCount = childrenIds.filter && childrenIds.filter( childId => treeNode?.children[childId]?.expanded).length || 0;
        return expandedChildrenCount;
    };

    const getMaxExpandedNodeLevel = (node) => {
        const expandedChildrenCount = getExpandedChildren(node);
        if (expandedChildrenCount < 1) {
            return node.level;
        }
        let maxExpandedLevel = node.level;
        const nodeChildren = node?.children || {};
        const nodeChildrenIds = Object.keys(nodeChildren);
        const expandedChildrenIds = nodeChildrenIds.filter(childId => node?.children[childId]?.expanded);
        for ( const expandedChildId of expandedChildrenIds) {
            const expandedChild = node?.children[expandedChildId];
            let expandedChildMaxDepth = getMaxExpandedNodeLevel(expandedChild);
            if (expandedChildMaxDepth > maxExpandedLevel) {
                maxExpandedLevel = expandedChildMaxDepth;
            }
        }
        return maxExpandedLevel;
    };

    const treeId = useUniqueId(id);

    const levelOneNodes = React.Children.toArray(children)
        .filter(child => child.type && child.type.displayName === 'Tree.Node')
        .map(child => React.cloneElement(child, {
            ...child?.props,
            level: 1,
            expansionEventBubbler: (toggledAtLevel, lastLevelExpanding, itemId, parentId, parentPath) => {
                updateTree({
                    toggledAtLevel,
                    lastLevelExpanding,
                    itemId,
                    parentId,
                    parentPath
                });
                const newExpansionDepth = getMaxExpandedNodeLevel(tree.root);
                setExpansionDepth(newExpansionDepth);
            },
            parentId: 'root',
            parentPath: 'root',
            selection,
            selectionPosition,
            treeId,
            treeActive: active
        }));

    const emptyTree = !levelOneNodes?.length > 0;


    const treeClasses = classnames(
        `${cssNamespace}-tree`,
        {
            [`expanded-level-${expansionDepth}`]: typeof expansionDepth === 'number' && expansionDepth > 0,
            [`${cssNamespace}-tree--active`]: active,
            [`${cssNamespace}-tree--compact`]: compact,
            [`${cssNamespace}-tree--no-border`]: noBorders,
            [`${cssNamespace}-tree--no-data`]: emptyTree
        },
        className
    );

    return (
        <ul
            ref={ref}
            {...props}
            className={treeClasses}
            id={treeId}
            role='tree'>
            {!emptyTree ?
                levelOneNodes
                : (
                    <li aria-level='1' className={classnames(`${cssNamespace}-tree__item`)}
                        role='treeitem'>
                        <div className={classnames(`${cssNamespace}-tree__item-container`)}>
                            <div className={classnames(`${cssNamespace}-tree__content`)}>
                                <span className={classnames(`${cssNamespace}-tree__text`)}>{emptyText}</span>
                            </div>
                        </div>
                    </li>)}
        </ul>);
});

Tree.displayName = 'Tree';

Tree.propTypes = {
    /** Set to **true** to make all `TreeNode`s have styles for interaction states (hover, selected, active).*/
    active: PropTypes.bool,
    /** React nodes to render within. Nest `TreeNode`s to create multiple levels in the `Tree`.*/
    children: PropTypes.node,
    /** Custom classes to add to Tree.*/
    className: PropTypes.string,
    /** Set to **true** to apply compact styles.*/
    compact: PropTypes.bool,
    /** Set to a **String** value to display when `Tree` is empty, i.e. has no children.*/
    emptyText: PropTypes.string,
    /** Set to a **String** value to use as id for root node which is an `<ul>`. If unset, a generate value will be used.*/
    id: PropTypes.string,
    /** Set to **true** to remove borders from all level 1 nodes.*/
    noBorders: PropTypes.bool,
    /** Set to `multi` to enable selecting multiple selection of `TreeNode`s with Checkboxes.
     *
     * Set to `single` to enable selecting only one of the `TreeNode`s at a time, with radio buttons.
    */
    selection: PropTypes.oneOf([
        'multi',
        'single'
    ]),
    /** Set to `left` to show selection control before node contents.
     *
     * Set to `right` to show selection control after node contents.
    */
    selectionPosition: PropTypes.oneOf([
        'left',
        'right'
    ])
};

Tree.defaultProps = {
    active: false,
    selectionPosition: 'left'
};

Tree.Node = TreeNode;

export default withStyles(Tree);
