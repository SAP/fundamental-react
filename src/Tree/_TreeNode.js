import Checkbox from '../Forms/Checkbox';
import classnamesBind from 'classnames/bind';
import FormRadioItem from '../Forms/FormRadioItem';
import Icon from '../Icon/Icon';
import PropTypes from 'prop-types';
import { TREE_NODE_HIGHLIGHTS } from '../utils/constants';
import useUniqueId from '../utils/useUniqueId';
import withStyles from '../utils/withStyles';
import React, { useState } from 'react';
import styles from 'fundamental-styles/dist/tree.css';

const classnames = classnamesBind.bind(styles);

const TreeNode = ({
    actions,
    active,
    children,
    className,
    cssNamespace,
    glyphsAfter,
    glyphsBefore,
    highlight,
    id,
    isNavigated,
    link,
    onExpandToggle,
    onSelectionChange,
    selectionProps,
    nodeData,
    wrapContent,
    ...rest
}) => {
    const [isExpanded, setExpanded] = useState(false);
    const [isSelected, setSelected] = useState(false);

    const { level } = rest;
    const nodeId = `${useUniqueId(id)}`;
    let nodePath = rest?.parentPath?.split?.('/') || ['root'];
    if (!nodePath?.includes?.(nodeId) && nodePath?.push) {
        nodePath.push(nodeId);
    }
    nodePath = nodePath?.join && nodePath.join('/');

    const nodeContentPayload = React.Children.toArray(children).filter(child => child?.type?.displayName !== 'Tree.Node');
    const nodeTextContent = React.Children.toArray(children).find(child => typeof child === 'string');


    const childrenLevel = level + 1;


    const nodeChildren = React.Children.toArray(children)
        .filter(child => child?.type?.displayName === 'Tree.Node')
        .map(child => React.cloneElement(child, {
            active,
            level: childrenLevel,
            expansionEventBubbler: rest?.expansionEventBubbler,
            parentId: nodeId,
            parentPath: nodePath,
            selection: rest.selection,
            selectionPosition: rest.selectionPosition,
            treeId: rest.treeId,
            treeActive: rest?.treeActive
        }));

    const handleButtonClick = (event) => {
        const toggledExpansionState = !isExpanded;
        rest?.expansionEventBubbler?.(level, toggledExpansionState, nodeId, rest?.parentId, nodePath);
        onExpandToggle?.(event, toggledExpansionState, nodeData);
        setExpanded(toggledExpansionState);
    };

    const validHighlight = () => TREE_NODE_HIGHLIGHTS.includes(highlight);

    const highlightModifier = () => {
        let modifier = '';
        if (validHighlight() && highlight !== 'default') {
            modifier = `--${highlight}`;
        }
        return modifier;
    };

    const hasNavigation = !!link?.trim();

    const nodeClasses = classnames(
        `${cssNamespace}-tree__item`,
        className
    );

    const containerClasses = classnames(
        `${cssNamespace}-tree__item-container`,
        {
            [`has-highlight-indicator${highlightModifier()}`]: validHighlight(),
            [`${cssNamespace}-tree__item-container--active`]: hasNavigation || active,
            'is-navigated': isNavigated,
            'is-selected': rest?.selection === 'multi' && isSelected
        }
    );

    const buttonClasses = classnames(
        `${cssNamespace}-tree__expander`,
        { 'is-expanded': !!isExpanded }
    );

    const contentClasses = classnames(
        `${cssNamespace}-tree__content`,
        { [`${cssNamespace}-tree__content--wrap`]: wrapContent },
        { 'has-navigation-indicator': hasNavigation && !rest?.treeActive }
    );

    const branchClasses = classnames(
        `${cssNamespace}-tree`,
        {
            'is-hidden': level > 1 && !isExpanded
        });

    const nodeContent = (
        <span className={classnames(`${cssNamespace}-tree__text`)}>
            {nodeContentPayload}
        </span>
    );

    let checkBox;
    let radioItem;
    const selectionControlId = `${nodeId}-selection-control`;

    if ( rest?.selection === 'multi' ) {
        const { onChange } = selectionProps || {};
        checkBox = (
            <Checkbox
                ariaLabel={nodeTextContent}
                {...selectionProps}
                id={selectionControlId}
                onChange={(event, checked) => {
                    onChange && onChange(event, checked);
                    onSelectionChange?.(event, checked, nodeData);
                    setSelected(checked);
                }} />
        );
    } else if (rest?.selection === 'single') {
        const { name } = selectionProps || {};
        const radioItemName = rest?.treeId;
        const { onChange } = selectionProps?.inputProps || {};

        radioItem = (
            <FormRadioItem
                children=''
                {...selectionProps}
                id={ selectionControlId }
                inputProps={{
                    'aria-label': nodeTextContent,
                    ...selectionProps?.inputProps,
                    onChange: (event) => {
                        onChange && onChange(event);
                        // always return true because radio item change is only triggered on selection not de-selection
                        onSelectionChange && onSelectionChange(event, true, nodeData);
                    }
                }}
                name={name ? name : radioItemName } />
        );
    }

    const { expansionEventBubbler, level: removeLevel, parentId, parentPath, selectionPosition, treeId, treeActive, ...otherProps } = rest;


    const renderIcons = (icons) => {
        return icons && icons?.length && icons?.map((icon, index) => (<Icon className={classnames(`${cssNamespace}-tree__icon`)} glyph={icon}
            key={index} />)) || '';
    };

    const buttonId = `${nodeId}-expansion-toggle-button`;
    const subTreeId = `${nodeId}-subtree-root`;

    return (
        <li
            {...otherProps}
            aria-expanded={isExpanded}
            aria-level={level}
            className={nodeClasses}
            id={nodeId}
            role='treeitem'>
            <div className={containerClasses}>
                {nodeChildren?.length > 0 &&
                    <button
                        aria-controls={nodeId}
                        aria-label={isExpanded ? 'collapse' : 'expand'}
                        aria-pressed={isExpanded}
                        className={buttonClasses}
                        id={buttonId}
                        onClick={handleButtonClick}>
                        <Icon ariaHidden className={classnames('fd-tree__icon')}
                            glyph={isExpanded ? 'navigation-down-arrow' : 'navigation-right-arrow'} />
                    </button>
                }
                {
                    renderIcons(glyphsBefore)
                }
                {
                    rest.selectionPosition === 'left' && rest?.selection === 'single' &&
                        radioItem
                }
                {
                    rest.selectionPosition === 'left' && rest?.selection === 'multi' &&
                        checkBox
                }
                {
                    hasNavigation ? (
                        <a className={contentClasses} href={link} >
                            {nodeContent}
                        </a>
                    ) : (
                        <div className={contentClasses}>
                            {nodeContent}
                        </div>
                    )
                }
                {
                    renderIcons(glyphsAfter)
                }
                {actions}
                {
                    rest.selectionPosition === 'right' && rest?.selection === 'single' &&
                        radioItem
                }
                {
                    rest.selectionPosition === 'right' && rest?.selection === 'multi' &&
                        checkBox
                }
            </div>
            {nodeChildren?.length > 0 &&
            <ul
                aria-hidden={!isExpanded}
                className={branchClasses}
                id={subTreeId}
                role='group'>
                {nodeChildren}
            </ul>
            }
        </li>
    );
};

TreeNode.displayName = 'Tree.Node';


TreeNode.propTypes = {
    /** React nodes to render as `TreeNode`s actions. Expecting `Button`s here.*/
    actions: PropTypes.node,
    /** Set to **true** to make this particular `TreeNode` have styles for interaction states (hover, selected, active).*/
    active: PropTypes.bool,
    /** React nodes to render within. Nest `TreeNode`s to create multiple levels in the `Tree`.*/
    children: PropTypes.node,
    /** Custom classes to add to TreeNode.*/
    className: PropTypes.string,
    /** An **Array of string** values to representing icon names to display before node contents.*/
    glyphsAfter: PropTypes.arrayOf(PropTypes.string),
    /** An **Array of string** values to representing icon names to display before after contents.*/
    glyphsBefore: PropTypes.arrayOf(PropTypes.string),
    /** Set to one of the valid values to highlight the row with the corresponding semantic color.*/
    highlight: PropTypes.oneOf([
        'default',
        'error',
        'success',
        'warning'
    ]),
    /** Set to a **String** value to use as id for this node which is a `<li>`. If unset, a generate value will be used.*/
    id: PropTypes.string,
    /** Set to **true** to show styles representing that the node has been navigate to.*/
    isNavigated: PropTypes.bool,
    /** Set to a **String**  value to use as href and make this node an `<a>` tag with appropriate styles.*/
    link: PropTypes.string,
    /** **Object** to be received in the `onExpandToggle` and `onSelectionChange` callbacks of this node.*/
    nodeData: PropTypes.object,
    /** Additional props to be spread to the selection control, if any.*/
    selectionProps: PropTypes.object,
    /** Set to **true** to make node contents wrap to next line(s).*/
    wrapContent: PropTypes.bool,
    /**
     * Callback function; triggered when a node's expansion button is clicked.
     * @param {SyntheticEvent} event - React's original SyntheticEvent.
     * @param {Boolean} expansion - Boolean representing if this event will expand TreeNode's subtree.
     * @param {Object} nodeData - TreeNode.nodeData
     * @returns {void}
    */
    onExpandToggle: PropTypes.func,
    /**
     * Callback function; triggered when a node is selected (if selection is enabled)
     *
     * * multi: selection state changes
     * * single: is selected
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent.
     * @param {Boolean} checked - Boolean representing if this event will select the TreeNode.
     * This is always true for single select Tree's because this event is only fired on radio selection.
     * @param {Object} nodeData - TreeNode.nodeData
    */
    onSelectionChange: PropTypes.func
};


export default withStyles(TreeNode);
