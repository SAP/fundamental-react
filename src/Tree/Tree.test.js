/* eslint-disable react/prop-types */
import { act } from 'react-test-renderer';
import { mount } from 'enzyme';
import React from 'react';
import {
    toHaveClass
} from '@testing-library/jest-dom/matchers';
import Tree from './Tree';
import { unmountComponentAtNode } from 'react-dom';

expect.extend({ toHaveClass });
let container = null;
beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() =>{
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

describe('<Tree />', () => {
    const tree = ({
        customNode = <Tree.Node> Level 7</Tree.Node>,
        treeProps = {},
        levelOneProps = {}
    } = {} ) => (
        <Tree {...treeProps}>
            <Tree.Node
                {...levelOneProps}
                id='myCustomLevel1'>
                Level 1
                <Tree.Node>
                    Level 2
                    <Tree.Node>
                        Level 3
                        <Tree.Node
                            id='myCustomLevel4'>
                            Level 4
                            <Tree.Node>
                                Level 5
                                <Tree.Node>
                                    Level 6
                                    {customNode}
                                </Tree.Node>
                            </Tree.Node>
                        </Tree.Node>
                    </Tree.Node>
                </Tree.Node>
            </Tree.Node>
            <Tree.Node>
                Level 1
                <Tree.Node>
                    Level 2
                    <Tree.Node>
                        Level 3
                        <Tree.Node>
                            Level 4
                            <Tree.Node>
                                Level 5
                                <Tree.Node>
                                    Level 6
                                    <Tree.Node>
                                        Level 7
                                    </Tree.Node>
                                </Tree.Node>
                            </Tree.Node>
                        </Tree.Node>
                    </Tree.Node>
                </Tree.Node>
            </Tree.Node>
        </Tree>
    );

    describe('Interactions', () => {
        test('Expand button should toggle subtree visibility', () => {
            const wrapper = mount(tree());
            const subTree = wrapper.find('#myCustomLevel1-subtree-root');
            expect(subTree.getDOMNode().getAttribute('aria-hidden')).toBe('true');

            const button = wrapper.find('#myCustomLevel1-expansion-toggle-button');
            act(() => {
                button.simulate('click');
            });
            expect(subTree.getDOMNode().getAttribute('aria-hidden')).toBe('false');
            act(() => {
                button.simulate('click');
            });
            expect(subTree.getDOMNode().getAttribute('aria-hidden')).toBe('true');
            wrapper.unmount();
        });

        test('Expand button onClick callback is called with correct parameters', () => {
            const mockHandler = jest.fn();
            const customNodeData = {
                level: 7,
                displayName: 'Level 4 + 1'
            };
            const wrapper = mount(tree({
                customNode: <Tree.Node
                    id='level7'
                    nodeData={customNodeData}
                    onExpandToggle={mockHandler}>
                    Level 7
                    <Tree.Node>
                        Level 8
                    </Tree.Node>
                </Tree.Node>
            }));

            const button = wrapper.find('#level7-expansion-toggle-button');
            act(() => {
                button.simulate('click');
            });
            expect(mockHandler).toHaveBeenCalledTimes(1);
            expect(mockHandler).toHaveBeenCalledWith(expect.anything(), true, customNodeData);
            act(() => {
                button.simulate('click');
            });
            expect(mockHandler).toHaveBeenCalledTimes(2);
            expect(mockHandler).toHaveBeenCalledWith(expect.anything(), false, customNodeData);
            wrapper.unmount();
        });

        test('TreeNodes should be multi-selectable in a multi select tree, selected container styles should apply', () => {
            const wrapper = mount(tree({
                customNode: <Tree.Node
                    id='level7'>
                    Level 7
                    <Tree.Node>
                        Level 8
                    </Tree.Node>
                </Tree.Node>,
                treeProps: {
                    selection: 'multi'
                }
            }));

            const levelOneCheckBox = wrapper.find('input#myCustomLevel1-selection-control');
            act(() => {
                levelOneCheckBox.simulate('change', { target: {
                    checked: true
                } });// simulate the real checked
            });
            expect(levelOneCheckBox.getDOMNode().getAttribute('aria-checked')).toBeTruthy();

            const levelOneContainer = wrapper.find('li#myCustomLevel1 > .fd-tree__item-container');
            expect(levelOneContainer.getDOMNode()).toHaveClass('is-selected');

            const level7CheckBox = wrapper.find('input#level7-selection-control');
            act(() => {
                level7CheckBox.simulate('change', { target: {
                    checked: true
                } }); // simulate the real checked
            });
            expect(level7CheckBox.getDOMNode().getAttribute('aria-checked')).toBeTruthy();

            const level7Container = wrapper.find('li#level7 > .fd-tree__item-container');
            expect(level7Container.getDOMNode()).toHaveClass('is-selected');

            wrapper.unmount();
        });

        test('should execute onSelectionChange callback with correct parameters in multi-select Tree', () => {
            const levelOneHandler = jest.fn();
            const levelOneData = {
                text: 'test'
            };
            const levelSevenHandler = jest.fn();
            const levelSevenData = {
                level: 7,
                displayName: 'Level 4 + 1'
            };
            const wrapper = mount(tree({
                customNode: <Tree.Node
                    id='level7'
                    nodeData={levelSevenData}
                    onSelectionChange={levelSevenHandler}>
                    Level 7
                    <Tree.Node>
                        Level 8
                    </Tree.Node>
                </Tree.Node>,
                treeProps: {
                    selection: 'multi'
                },
                levelOneProps: {
                    onSelectionChange: levelOneHandler,
                    nodeData: levelOneData
                }
            }));
            const levelOneCheckBox = wrapper.find('input#myCustomLevel1-selection-control');
            act(() => {
                levelOneCheckBox.simulate('change', { target: {
                    checked: true
                } });
            });
            expect(levelOneHandler).toHaveBeenCalledTimes(1);
            expect(levelOneHandler).toHaveBeenCalledWith(expect.anything(), true, levelOneData);

            const level7CheckBox = wrapper.find('input#level7-selection-control');
            act(() => {
                level7CheckBox.simulate('change', { target: {
                    checked: true
                } });
            });
            expect(levelSevenHandler).toHaveBeenCalledTimes(1);
            expect(levelSevenHandler).toHaveBeenCalledWith(expect.anything(), true, levelSevenData);

            wrapper.unmount();
        });

        test('should execute onSelectionChange callback with correct parameters in single-select Tree', () => {
            const radioOneHandler = jest.fn();
            const radioOneData = {
                text: 'test'
            };
            const radioTwoHandler = jest.fn();
            const radioTwoData = {
                level: 7,
                displayName: 'Level 4 + 1'
            };
            const wrapper = mount(
                <Tree id='radioTesting' selection='single'>
                    <Tree.Node id='radio1' nodeData={radioOneData}
                        onSelectionChange={radioOneHandler}>
                        Level 1
                    </Tree.Node>
                    <Tree.Node id='radio2' nodeData={radioTwoData}
                        onSelectionChange={radioTwoHandler}>
                        Level 1
                    </Tree.Node>
                </Tree>
            );
            const levelOneCheckBox = wrapper.find('input#radio1-selection-control');
            act(() => {
                levelOneCheckBox.simulate('change');
            });
            expect(radioOneHandler).toHaveBeenCalledTimes(1);
            expect(radioOneHandler).toHaveBeenCalledWith(expect.anything(), true, radioOneData);

            const levelTwoCheckBox = wrapper.find('input#radio2-selection-control');
            act(() => {
                levelTwoCheckBox.simulate('change');
            });
            expect(radioTwoHandler).toHaveBeenCalledTimes(1);
            expect(radioTwoHandler).toHaveBeenCalledWith(expect.anything(), true, radioTwoData);
            wrapper.unmount();
        });
    });

    describe('Props spreading', () => {
        test('should allow props to be spread on the Tree component', () => {
            const wrapper = mount(tree({
                treeProps: {
                    'data-sample': 'free-samples',
                    id: 'crazyTree',
                    className: 'crazy-styles'
                }
            }));
            const treeRoot = wrapper.find('ul#crazyTree');
            expect(
                treeRoot.getDOMNode().getAttribute('data-sample')
            ).toBe('free-samples');

            expect(
                treeRoot.getDOMNode()
            ).toHaveClass('crazy-styles');
            wrapper.unmount();
        });
    });
});
