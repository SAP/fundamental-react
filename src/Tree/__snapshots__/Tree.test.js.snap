// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tree /> create tree component 1`] = `
<div>
  <div
    className="fd-tree fd-tree--header"
  >
    <div
      className="fd-tree__row fd-tree__row--header"
    >
      <div
        className="fd-tree__col fd-tree__col--control"
      >
        <button
          aria-label="expand"
          aria-pressed={false}
          className="fd-tree__control "
          onClick={[Function]}
        />
        Column Header
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 1 
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 2
      </div>
      <div
        className="fd-tree__col"
      >
        Status
      </div>
    </div>
  </div>
  <ul
    className="fd-tree"
    role="tree"
  >
    <li
      aria-expanded="true"
      className="fd-tree__item"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <button
            aria-controls="inYUX852"
            aria-label="expand"
            className="fd-tree__control"
            onClick={[Function]}
          />
          First Level
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          INACTIVE
        </div>
        <div
          className="fd-dropdown"
        />
      </div>
    </li>
    <li
      aria-expanded="true"
      className="fd-tree__item"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <button
            aria-controls="inYUX852"
            aria-label="expand"
            className="fd-tree__control"
            onClick={[Function]}
          />
          Row 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          DEFAULT
        </div>
        <div
          className="fd-dropdown"
        />
      </div>
    </li>
    <li
      aria-expanded="true"
      className="fd-tree__item"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          Row 3
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          INACTIVE
        </div>
        <div
          className="fd-dropdown"
        />
      </div>
    </li>
    <li
      aria-expanded="true"
      className="fd-tree__item"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <button
            aria-controls="inYUX852"
            aria-label="expand"
            className="fd-tree__control"
            onClick={[Function]}
          />
          <a
            className="fd-has-font-weight-semi"
            href="http://me.com"
          >
            http://me.com
          </a>
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          INACTIVE
        </div>
        <div
          className="fd-dropdown"
        />
      </div>
    </li>
  </ul>
</div>
`;

exports[`<Tree /> create tree component 2`] = `
<div>
  <div
    className="fd-tree fd-tree--header"
  >
    <div
      className="fd-tree__row fd-tree__row--header"
    >
      <div
        className="fd-tree__col fd-tree__col--control"
      >
        <button
          aria-label="expand"
          aria-pressed={false}
          className="fd-tree__control "
          onClick={[Function]}
        />
        Column Header
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 1 
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 2
      </div>
      <div
        className="fd-tree__col"
      >
        Status
      </div>
    </div>
  </div>
  <ul
    className="fd-tree"
    role="tree"
  >
    <li
      aria-expanded="true"
      className="fd-tree__item"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <button
            aria-controls="inYUX852"
            aria-label="expand"
            className="fd-tree__control"
            onClick={[Function]}
          />
          <a
            className="fd-has-font-weight-semi"
            href="http://me.com"
          >
            First Level
          </a>
        </div>
        <div
          className="fd-tree__col"
        >
           
        </div>
        <div
          className="fd-tree__col"
        >
           
        </div>
        <div
          className="fd-tree__col"
        >
           
        </div>
        <div
          className="fd-dropdown"
        />
      </div>
    </li>
    <li
      aria-expanded="true"
      className="fd-tree__item"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <button
            aria-controls="inYUX852"
            aria-label="expand"
            className="fd-tree__control"
            onClick={[Function]}
          />
          Second Level
        </div>
        <div
          className="fd-tree__col"
        >
           
        </div>
        <div
          className="fd-tree__col"
        >
           
        </div>
        <div
          className="fd-tree__col"
        >
           
        </div>
        <div
          className="fd-dropdown"
        />
      </div>
    </li>
  </ul>
</div>
`;
